File: src/main/java/link/infra/packwiz/installer/UpdateManager.java
package link.infra.packwiz.installer;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonIOException;
import com.google.gson.JsonSyntaxException;
import com.google.gson.annotations.SerializedName;
import com.moandjiezana.toml.Toml;
import link.infra.packwiz.installer.metadata.IndexFile;
import link.infra.packwiz.installer.metadata.ManifestFile;
import link.infra.packwiz.installer.metadata.PackFile;
import link.infra.packwiz.installer.metadata.hash.GeneralHashingSource;
import link.infra.packwiz.installer.metadata.hash.Hash;
import link.infra.packwiz.installer.metadata.hash.HashUtils;
import link.infra.packwiz.installer.request.HandlerManager;
import link.infra.packwiz.installer.ui.IExceptionDetails;
import link.infra.packwiz.installer.ui.IUserInterface;
import link.infra.packwiz.installer.ui.InstallProgress;
import okio.Okio;
import okio.Source;

import java.io.*;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class UpdateManager {

	private final Options opts;
	public final IUserInterface ui;
	private boolean cancelled;
	private boolean cancelledStartGame = false;

	public static class Options {
		URI downloadURI = null;
		String manifestFile = "packwiz.json"; // TODO: make configurable
		String packFolder = ".";
		Side side = Side.CLIENT;

		public enum Side {
			@SerializedName("client")
			CLIENT("client"),
			@SerializedName("server")
			SERVER("server"),
			@SerializedName("both")
			BOTH("both", new Side[] { CLIENT, SERVER });

			private final String sideName;
			private final Side[] depSides;

			Side(String sideName) {
				this.sideName = sideName.toLowerCase();
				this.depSides = null;
			}

			Side(String sideName, Side[] depSides) {
				this.sideName = sideName.toLowerCase();
				this.depSides = depSides;
			}

			@Override
			public String toString() {
				return this.sideName;
			}

			public boolean hasSide(Side tSide) {
				if (this.equals(tSide)) {
					return true;
				}
				if (this.depSides != null) {
					for (Side depSide : this.depSides) {
						if (depSide.equals(tSide)) {
							return true;
						}
					}
				}
				return false;
			}

			public static Side from(String name) {
				String lowerName = name.toLowerCase();
				for (Side side : Side.values()) {
					if (side.sideName.equals(lowerName)) {
						return side;
					}
				}
				return null;
			}
		}
	}

	UpdateManager(Options opts, IUserInterface ui) {
		this.opts = opts;
		this.ui = ui;
		this.start();
	}

	private void start() {
		this.checkOptions();

		ui.submitProgress(new InstallProgress("Loading manifest file..."));
		Gson gson = new GsonBuilder().registerTypeAdapter(Hash.class, new Hash.TypeHandler()).create();
		ManifestFile manifest;
		try {
			manifest = gson.fromJson(new FileReader(Paths.get(opts.packFolder, opts.manifestFile).toString()),
					ManifestFile.class);
		} catch (FileNotFoundException e) {
			manifest = new ManifestFile();
		} catch (JsonSyntaxException | JsonIOException e) {
			ui.handleExceptionAndExit(e);
			return;
		}

		ui.submitProgress(new InstallProgress("Loading pack file..."));
		GeneralHashingSource packFileSource;
		try {
			Source src = HandlerManager.getFileSource(opts.downloadURI);
			packFileSource = HashUtils.getHasher("sha256").getHashingSource(src);
		} catch (Exception e) {
			// TODO: still launch the game if updating doesn't work?
			// TODO: ask user if they want to launch the game, exit(1) if they don't
			ui.handleExceptionAndExit(e);
			return;
		}
		PackFile pf;
		try {
			pf = new Toml().read(Okio.buffer(packFileSource).inputStream()).to(PackFile.class);
		} catch (IllegalStateException e) {
			ui.handleExceptionAndExit(e);
			return;
		}

		ui.submitProgress(new InstallProgress("Checking local files..."));

		// Invalidation checking must be done here, as it must happen before pack/index hashes are checked
		List<URI> invalidatedUris = new ArrayList<>();
		if (manifest.cachedFiles != null) {
			for (Map.Entry<URI, ManifestFile.File> entry : manifest.cachedFiles.entrySet()) {
				boolean invalid = false;
				// if isn't optional, or is optional but optionValue == true
				if (!entry.getValue().isOptional || entry.getValue().optionValue) {
					if (entry.getValue().cachedLocation != null) {
						if (!Files.exists(Paths.get(opts.packFolder, entry.getValue().cachedLocation))) {
							invalid = true;
						}
					} else {
						// if cachedLocation == null, should probably be installed!!
						invalid = true;
					}
				}
				if (invalid) {
					URI fileUri = entry.getKey();
					System.out.println("File " + fileUri.toString() + " invalidated, marked for redownloading");
					invalidatedUris.add(fileUri);
				}
			}
		}

		if (manifest.packFileHash != null && packFileSource.hashIsEqual(manifest.packFileHash) && invalidatedUris.isEmpty()) {
			System.out.println("Modpack is already up to date!");
			// todo: --force?
			return;
		}

		System.out.println("Modpack name: " + pf.name);

		try {
			// This is badly written, I'll probably heavily refactor it at some point
			processIndex(HandlerManager.getNewLoc(opts.downloadURI, pf.index.file),
					HashUtils.getHash(pf.index.hashFormat, pf.index.hash), pf.index.hashFormat, manifest, invalidatedUris);
		} catch (Exception e1) {
			ui.handleExceptionAndExit(e1);
		}

		if (cancelled) {
			System.out.println("Update cancelled by user!");
			System.exit(1);
			return;
		} else if (cancelledStartGame) {
			System.out.println("Update cancelled by user! Continuing to start game...");
			System.exit(0);
			return;
		}

		// TODO: update MMC params, java args etc

		manifest.packFileHash = packFileSource.getHash();
		manifest.cachedSide = opts.side;
		try (Writer writer = new FileWriter(Paths.get(opts.packFolder, opts.manifestFile).toString())) {
			gson.toJson(manifest, writer);
		} catch (IOException e) {
			// TODO: add message?
			ui.handleException(e);
		}

	}

	private void checkOptions() {
		// TODO: implement
	}

	private void processIndex(URI indexUri, Hash indexHash, String hashFormat, ManifestFile manifest, List<URI> invalidatedUris) {
		if (manifest.indexFileHash != null && manifest.indexFileHash.equals(indexHash) && invalidatedUris.isEmpty()) {
			System.out.println("Modpack files are already up to date!");
			return;
		}
		manifest.indexFileHash = indexHash;

		GeneralHashingSource indexFileSource;
		try {
			Source src = HandlerManager.getFileSource(indexUri);
			indexFileSource = HashUtils.getHasher(hashFormat).getHashingSource(src);
		} catch (Exception e) {
			// TODO: still launch the game if updating doesn't work?
			// TODO: ask user if they want to launch the game, exit(1) if they don't
			ui.handleExceptionAndExit(e);
			return;
		}
		IndexFile indexFile;
		try {
			indexFile = new Toml().read(Okio.buffer(indexFileSource).inputStream()).to(IndexFile.class);
		} catch (IllegalStateException e) {
			ui.handleExceptionAndExit(e);
			return;
		}

		if (!indexFileSource.hashIsEqual(indexHash)) {
			// TODO: throw exception
			System.out.println("I was meant to put an error message here but I'll do that later");
			return;
		}

		if (manifest.cachedFiles == null) {
			manifest.cachedFiles = new HashMap<>();
		}

		ui.submitProgress(new InstallProgress("Checking local files..."));
		Iterator<Map.Entry<URI, ManifestFile.File>> it = manifest.cachedFiles.entrySet().iterator();
		while (it.hasNext()) {
			Map.Entry<URI, ManifestFile.File> entry = it.next();
			if (entry.getValue().cachedLocation != null) {
				boolean alreadyDeleted = false;
				// Delete if option value has been set to false
				if (entry.getValue().isOptional && !entry.getValue().optionValue) {
					try {
						Files.deleteIfExists(Paths.get(opts.packFolder, entry.getValue().cachedLocation));
					} catch (IOException e) {
						// TODO: should this be shown to the user in some way?
						e.printStackTrace();
					}
					// Set to null, as it doesn't exist anymore
					entry.getValue().cachedLocation = null;
					alreadyDeleted = true;
				}
				if (indexFile.files.stream().noneMatch(f -> f.file.equals(entry.getKey()))) {
					// File has been removed from the index
					if (!alreadyDeleted) {
						try {
							Files.deleteIfExists(Paths.get(opts.packFolder, entry.getValue().cachedLocation));
						} catch (IOException e) {
							// TODO: should this be shown to the user in some way?
							e.printStackTrace();
						}
					}
					it.remove();
				}
			}
		}
		ui.submitProgress(new InstallProgress("Comparing new files..."));

		// TODO: progress bar, parallelify
		List<DownloadTask> tasks = DownloadTask.createTasksFromIndex(indexFile, indexFile.hashFormat, opts.side);
		// If the side changes, invalidate EVERYTHING just in case
		// Might not be needed, but done just to be safe
		boolean invalidateAll = !opts.side.equals(manifest.cachedSide);
		tasks.forEach(f -> {
			// TODO: should linkedfile be checked as well? should this be done in the download section?
			if (invalidateAll) {
				f.invalidate();
			} else if (invalidatedUris.contains(f.metadata.file)) {
				f.invalidate();
			}
			ManifestFile.File file = manifest.cachedFiles.get(f.metadata.file);
			if (file != null) {
				// Ensure the file can be reverted later if necessary - the DownloadTask modifies the file so if it fails we need the old version back
				file.backup();
			}
			// If it is null, the DownloadTask will make a new empty cachedFile
			f.updateFromCache(file);
		});
		tasks.forEach(f -> f.downloadMetadata(indexFile, indexUri));

		List<IExceptionDetails> failedTasks = tasks.stream().filter(t -> t.getException() != null).collect(Collectors.toList());
		if (failedTasks.size() > 0) {
			IExceptionDetails.ExceptionListResult exceptionListResult;
			try {
				exceptionListResult = ui.showExceptions(failedTasks, tasks.size(), true).get();
			} catch (InterruptedException | ExecutionException e) {
				// Interrupted means cancelled???
				ui.handleExceptionAndExit(e);
				return;
			}
			switch (exceptionListResult) {
				case CONTINUE:
					break;
				case CANCEL:
					cancelled = true;
					return;
				case IGNORE:
					cancelledStartGame = true;
					return;
			}
		}

		List<DownloadTask> nonFailedFirstTasks = tasks.stream().filter(t -> t.getException() == null).collect(Collectors.toList());
		List<DownloadTask> optionTasks = nonFailedFirstTasks.stream().filter(DownloadTask::correctSide).filter(DownloadTask::isOptional).collect(Collectors.toList());
		// If options changed, present all options again
		if (optionTasks.stream().anyMatch(DownloadTask::isNewOptional)) {
			// new ArrayList is requires so it's an IOptionDetails rather than a DownloadTask list
			Future<Boolean> cancelledResult = ui.showOptions(new ArrayList<>(optionTasks));
			try {
				if (cancelledResult.get()) {
					cancelled = true;
					// TODO: Should the UI be closed somehow??
					return;
				}
			} catch (InterruptedException | ExecutionException e) {
				// Interrupted means cancelled???
				ui.handleExceptionAndExit(e);
			}
		}

		// TODO: different thread pool type?
		ExecutorService threadPool = Executors.newFixedThreadPool(10);
		CompletionService<DownloadTask> completionService = new ExecutorCompletionService<>(threadPool);

		tasks.forEach(t -> completionService.submit(() -> {
			t.download(opts.packFolder, indexUri);
			return t;
		}));

		for (int i = 0; i < tasks.size(); i++) {
			DownloadTask task;
			try {
				task = completionService.take().get();
			} catch (InterruptedException | ExecutionException e) {
				ui.handleException(e);
				task = null;
			}
			// Update manifest - If there were no errors cachedFile has already been modified in place (good old pass by reference)
			if (task != null) {
				if (task.getException() != null) {
					ManifestFile.File file = task.cachedFile.getRevert();
					if (file != null) {
						manifest.cachedFiles.put(task.metadata.file, file);
					}
				} else {
					// idiot, if it wasn't there in the first place it won't magically appear there
					manifest.cachedFiles.putIfAbsent(task.metadata.file, task.cachedFile);
				}
			}

			String progress;
			if (task != null) {
				if (task.getException() != null) {
					progress = "Failed to download " + task.metadata.getName() + ": " + task.getException().getMessage();
					task.getException().printStackTrace();
				} else {
					// TODO: should this be revised for tasks that didn't actually download it?
					progress = "Downloaded " + task.metadata.getName();
				}
			} else {
				progress = "Failed to download, unknown reason";
			}
			ui.submitProgress(new InstallProgress(progress, i + 1, tasks.size()));
		}

		List<IExceptionDetails> failedTasks2ElectricBoogaloo = nonFailedFirstTasks.stream().filter(t -> t.getException() != null).collect(Collectors.toList());
		if (failedTasks2ElectricBoogaloo.size() > 0) {
			IExceptionDetails.ExceptionListResult exceptionListResult;
			try {
				exceptionListResult = ui.showExceptions(failedTasks2ElectricBoogaloo, tasks.size(), false).get();
			} catch (InterruptedException | ExecutionException e) {
				// Interrupted means cancelled???
				ui.handleExceptionAndExit(e);
				return;
			}
			switch (exceptionListResult) {
				case CONTINUE:
					break;
				case CANCEL:
					cancelled = true;
					return;
				case IGNORE:
					cancelledStartGame = true;
			}
		}
	}
}

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt
package link.infra.packwiz.installer

import cc.ekblad.toml.decode
import com.google.gson.GsonBuilder
import com.google.gson.JsonIOException
import com.google.gson.JsonSyntaxException
import link.infra.packwiz.installer.DownloadTask.Companion.createTasksFromIndex
import link.infra.packwiz.installer.metadata.DownloadMode
import link.infra.packwiz.installer.metadata.IndexFile
import link.infra.packwiz.installer.metadata.ManifestFile
import link.infra.packwiz.installer.metadata.PackFile
import link.infra.packwiz.installer.metadata.curseforge.resolveCfMetadata
import link.infra.packwiz.installer.metadata.hash.Hash
import link.infra.packwiz.installer.metadata.hash.HashFormat
import link.infra.packwiz.installer.request.RequestException
import link.infra.packwiz.installer.target.ClientHolder
import link.infra.packwiz.installer.target.Side
import link.infra.packwiz.installer.target.path.PackwizFilePath
import link.infra.packwiz.installer.target.path.PackwizPath
import link.infra.packwiz.installer.ui.IUserInterface
import link.infra.packwiz.installer.ui.IUserInterface.CancellationResult
import link.infra.packwiz.installer.ui.IUserInterface.ExceptionListResult
import link.infra.packwiz.installer.ui.data.InstallProgress
import link.infra.packwiz.installer.util.Log
import okio.buffer
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util.concurrent.CompletionService
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorCompletionService
import java.util.concurrent.Executors
import kotlin.system.exitProcess

class UpdateManager internal constructor(private val opts: Options, val ui: IUserInterface) {
	private var cancelled = false
	private var cancelledStartGame = false
	private var errorsOccurred = false

	init {
		start()
	}

	data class Options(
		val packFile: PackwizPath<*>,
		val manifestFile: PackwizFilePath,
		val packFolder: PackwizFilePath,
		val multimcFolder: PackwizFilePath,
		val side: Side,
		val timeout: Long,
	)

	// TODO: make this return a value based on results?
	private fun start() {
		val clientHolder = ClientHolder()
		ui.cancelCallback = {
			clientHolder.close()
		}

		ui.submitProgress(InstallProgress(Msgs.umLoadingManifest()))
		val gson = GsonBuilder()
			.registerTypeAdapter(Hash::class.java, Hash.TypeHandler())
			.registerTypeAdapter(PackwizFilePath::class.java, PackwizPath.adapterRelativeTo(opts.packFolder))
			.enableComplexMapKeySerialization()
			.create()
		val manifest = try {
			// TODO: kotlinx.serialisation?
			InputStreamReader(opts.manifestFile.source(clientHolder).inputStream(), StandardCharsets.UTF_8).use { reader ->
				gson.fromJson(reader, ManifestFile::class.java)
			}
		} catch (e: RequestException.Response.File.FileNotFound) {
			ui.firstInstall = true
			ManifestFile()
		} catch (e: JsonSyntaxException) {
			ui.showErrorAndExit(Msgs.umInvalidManifestSyntax(opts.manifestFile), e)
		} catch (e: JsonIOException) {
			ui.showErrorAndExit(Msgs.umInvalidManifestIO(opts.manifestFile), e)
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			handleCancellation()
		}

		ui.submitProgress(InstallProgress(Msgs.umLoadingPack()))
		val packFileSource = try {
			val src = opts.packFile.source(clientHolder)
			HashFormat.SHA256.source(src)
		} catch (e: Exception) {
			// TODO: ensure suppressed/caused exceptions are shown?
			ui.showErrorAndExit(Msgs.umFailedDownloadPack(), e)
		}
		val pf = packFileSource.buffer().use {
			try {
				PackFile.mapper(opts.packFile).decode<PackFile>(it.inputStream())
			} catch (e: IllegalStateException) {
				ui.showErrorAndExit(Msgs.umFailedParsePack(), e)
			}
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			handleCancellation()
		}

		// Launcher checks
		val lu = LauncherUtils(opts, ui)

		// MultiMC MC and loader version checker
		ui.submitProgress(InstallProgress(Msgs.umLoadingMultiMC()))
		try {
			when (lu.handleMultiMC(pf, gson)) {
				LauncherUtils.LauncherStatus.CANCELLED -> cancelled = true
				LauncherUtils.LauncherStatus.NOT_FOUND -> Log.info(Msgs.umNoMultiMC())
				else -> {}
			}
			handleCancellation()
		} catch (e: Exception) {
			ui.showErrorAndExit(e.message!!, e)
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			handleCancellation()
		}

		ui.submitProgress(InstallProgress(Msgs.umCheckingLocalFiles()))

		// If the side changes, invalidate EVERYTHING (even when the index hasn't changed)
		val invalidateAll = opts.side != manifest.cachedSide
		val invalidatedUris: MutableList<PackwizFilePath> = ArrayList()
		if (!invalidateAll) {
			// Invalidation checking must be done here, as it must happen before pack/index hashes are checked
			for ((fileUri, file) in manifest.cachedFiles) {
				// ignore onlyOtherSide files
				if (file.onlyOtherSide) {
					continue
				}

				var invalid = false
				// if isn't optional, or is optional but optionValue == true
				if (!file.isOptional || file.optionValue) {
					if (file.cachedLocation != null) {
						if (!file.cachedLocation!!.nioPath.toFile().exists()) {
							invalid = true
						}
					} else {
						// if cachedLocation == null, should probably be installed!!
						invalid = true
					}
				}
				if (invalid) {
					Log.info(Msgs.umInvalidateLocalFile(fileUri.filename))
					invalidatedUris.add(fileUri)
				}
			}

			if (manifest.packFileHash?.let { it == packFileSource.hash } == true && invalidatedUris.isEmpty()) {
				// todo: --force?
				ui.submitProgress(InstallProgress(Msgs.umModpackUpToDate(), 1, 1))
				if (manifest.cachedFiles.any { it.value.isOptional }) {
					ui.awaitOptionalButton(false, opts.timeout)
				}
				if (!ui.optionsButtonPressed) {
					return
				}
			}
		}

		Log.info(Msgs.umModpackName(pf.name))

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			handleCancellation()
		}
		try {
			processIndex(
				pf.index.file,
				pf.index.hashFormat.fromString(pf.index.hash),
				pf.index.hashFormat,
				manifest,
				invalidatedUris,
				invalidateAll,
				clientHolder
			)
		} catch (e1: Exception) {
			ui.showErrorAndExit(Msgs.umInvalidIndex(), e1)
		}

		handleCancellation()


		// If there were errors, don't write the manifest/index hashes, to ensure they are rechecked later
		if (errorsOccurred) {
			manifest.indexFileHash = null
			manifest.packFileHash = null
		} else {
			manifest.packFileHash = packFileSource.hash
		}

		manifest.cachedSide = opts.side
		try {
			Files.newBufferedWriter(opts.manifestFile.nioPath, StandardCharsets.UTF_8).use { writer -> gson.toJson(manifest, writer) }
		} catch (e: IOException) {
			ui.showErrorAndExit(Msgs.umFailedSaveLocalManifest(), e)
		}
	}

	private fun processIndex(indexUri: PackwizPath<*>, indexHash: Hash<*>, hashFormat: HashFormat<*>, manifest: ManifestFile, invalidatedFiles: List<PackwizFilePath>, invalidateAll: Boolean, clientHolder: ClientHolder) {
		if (!invalidateAll) {
			if (manifest.indexFileHash == indexHash && invalidatedFiles.isEmpty()) {
				ui.submitProgress(InstallProgress(Msgs.umModpackFilesUpToDate(), 1, 1))
				if (manifest.cachedFiles.any { it.value.isOptional }) {
					ui.awaitOptionalButton(false, opts.timeout)
				}
				if (!ui.optionsButtonPressed) {
					return
				}
				if (ui.cancelButtonPressed) {
					showCancellationDialog()
					return
				}
			}
		}
		manifest.indexFileHash = indexHash

		val indexFileSource = try {
			val src = indexUri.source(clientHolder)
			hashFormat.source(src)
		} catch (e: Exception) {
			ui.showErrorAndExit(Msgs.umFailedDownloadIndex(), e)
		}

		val indexFile = try {
			IndexFile.mapper(indexUri).decode<IndexFile>(indexFileSource.buffer().inputStream())
		} catch (e: IllegalStateException) {
			ui.showErrorAndExit(Msgs.umFailedParseIndex(), e)
		}
		if (indexHash != indexFileSource.hash) {
			ui.showErrorAndExit(Msgs.umInvalidIndexHash())
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			return
		}

		ui.submitProgress(InstallProgress(Msgs.umCheckingLocalFiles()))
		val it: MutableIterator<Map.Entry<PackwizFilePath, ManifestFile.File>> = manifest.cachedFiles.entries.iterator()
		while (it.hasNext()) {
			val (uri, file) = it.next()
			if (file.cachedLocation != null) {
				if (indexFile.files.none { it.file.rebase(opts.packFolder) == uri }) { // File has been removed from the index
					try {
						Files.deleteIfExists(file.cachedLocation!!.nioPath)
					} catch (e: IOException) {
						Log.warn(Msgs.umFailedDeleteFile(), e)
					}
					Log.info(Msgs.umDeletedFile(file.cachedLocation!!.filename))
					it.remove()
				}
			}
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			return
		}
		ui.submitProgress(InstallProgress(Msgs.umComparingNewFiles()))

		// TODO: progress bar?
		if (indexFile.files.isEmpty()) {
			Log.warn(Msgs.umIndexEmpty())
		}
		val tasks = createTasksFromIndex(indexFile, opts.side)
		if (invalidateAll) {
			Log.info(Msgs.umSideChanged())
		}
		tasks.forEach{ f ->
			// TODO: should linkedfile be checked as well? should this be done in the download section?
			if (invalidateAll) {
				f.invalidate()
			} else if (invalidatedFiles.contains(f.metadata.file.rebase(opts.packFolder))) {
				f.invalidate()
			}
			val file = manifest.cachedFiles[f.metadata.file.rebase(opts.packFolder)]
			// Ensure the file can be reverted later if necessary - the DownloadTask modifies the file so if it fails we need the old version back
			file?.backup()
			// If it is null, the DownloadTask will make a new empty cachedFile
			f.updateFromCache(file)
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			return
		}

		// Let's hope downloadMetadata is a pure function!!!
		tasks.parallelStream().forEach { f -> f.downloadMetadata(clientHolder) }

		val failedTaskDetails = tasks.asSequence().map(DownloadTask::exceptionDetails).filterNotNull().toList()
		if (failedTaskDetails.isNotEmpty()) {
			errorsOccurred = true
			when (ui.showExceptions(failedTaskDetails, tasks.size, true)) {
				ExceptionListResult.CONTINUE -> {}
				ExceptionListResult.CANCEL -> {
					cancelled = true
					return
				}
				ExceptionListResult.IGNORE -> {
					cancelledStartGame = true
					return
				}
			}
		}

		if (ui.cancelButtonPressed) {
			showCancellationDialog()
			return
		}

		// TODO: task failed function?
		tasks.removeAll { it.failed() }
		val optionTasks = tasks.filter(DownloadTask::correctSide).filter(DownloadTask::isOptional).toList()
		val optionsChanged = optionTasks.any(DownloadTask::isNewOptional)
		if (optionTasks.isNotEmpty() && !optionsChanged) {
			if (!ui.optionsButtonPressed) {
				// TODO: this is so ugly
				ui.submitProgress(InstallProgress(Msgs.umReconfigureOptionalQuestion(), 0,1))
				ui.awaitOptionalButton(true, opts.timeout)
				if (ui.cancelButtonPressed) {
					showCancellationDialog()
					return
				}
			}
		}
		// If options changed, present all options again
		if (ui.optionsButtonPressed || optionsChanged) {
			// new ArrayList is required so it's an IOptionDetails rather than a DownloadTask list
			if (ui.showOptions(ArrayList(optionTasks))) {
				cancelled = true
				handleCancellation()
			}
		}
		// TODO: keep this enabled? then apply changes after download process?
		ui.disableOptionsButton(optionTasks.isNotEmpty())

		while (true) {
			when (validateAndResolve(tasks, clientHolder)) {
				ResolveResult.RETRY -> {}
				ResolveResult.QUIT -> return
				ResolveResult.SUCCESS -> break
			}
		}

		// TODO: different thread pool type?
		val threadPool = Executors.newFixedThreadPool(10)
		val completionService: CompletionService<DownloadTask> = ExecutorCompletionService(threadPool)
		tasks.forEach { t ->
			completionService.submit {
				t.download(opts.packFolder, clientHolder)
				t
			}
		}
		for (i in tasks.indices) {
			val task: DownloadTask = try {
				completionService.take().get()
			} catch (e: InterruptedException) {
				ui.showErrorAndExit(Msgs.umInterruptedDownloadTask(), e)
			} catch (e: ExecutionException) {
				ui.showErrorAndExit(Msgs.umFailedDownloadTask(), e)
			}
			// Update manifest - If there were no errors cachedFile has already been modified in place (good old pass by reference)
			task.cachedFile?.let { file ->
				if (task.failed()) {
					val oldFile = file.revert
					if (oldFile != null) {
						manifest.cachedFiles.putIfAbsent(task.metadata.file.rebase(opts.packFolder), oldFile)
					} else { null }
				} else {
					manifest.cachedFiles.putIfAbsent(task.metadata.file.rebase(opts.packFolder), file)
				}
			}

			val exDetails = task.exceptionDetails
			val progress = if (exDetails != null) {
				Msgs.umFailedDownload(exDetails.name, exDetails.exception.message ?: "")
			} else {
				when (task.completionStatus) {
					DownloadTask.CompletionStatus.INCOMPLETE -> Msgs.umTaskStatePending(task.name)
					DownloadTask.CompletionStatus.DOWNLOADED -> Msgs.umTaskStateDownloaded(task.name)
					DownloadTask.CompletionStatus.ALREADY_EXISTS_CACHED -> Msgs.unTaskStateAlreadyExistsCached(task.name)
					DownloadTask.CompletionStatus.ALREADY_EXISTS_VALIDATED -> Msgs.umTaskStateAlreadyExistsValidated(task.name)
					DownloadTask.CompletionStatus.SKIPPED_DISABLED -> Msgs.umTaskStateSkippedDisabled(task.name)
					DownloadTask.CompletionStatus.SKIPPED_WRONG_SIDE -> Msgs.umTaskStateSkippedWrongSide(task.name)
					DownloadTask.CompletionStatus.DELETED_DISABLED -> Msgs.umTaskStateDeletedDisabled(task.name)
					DownloadTask.CompletionStatus.DELETED_WRONG_SIDE -> Msgs.umTaskStateDeletedWrongSide(task.name)
				}
			}
			ui.submitProgress(InstallProgress(progress, i + 1, tasks.size))

			if (ui.cancelButtonPressed) { // Stop all tasks, don't launch the game (it's in an invalid state!)
				// TODO: close client holder in more places?
				clientHolder.close()
				threadPool.shutdown()
				cancelled = true
				return
			}
		}

		// Shut down the thread pool when the update is done
		threadPool.shutdown()

		val failedTasks2ElectricBoogaloo = tasks.asSequence().map(DownloadTask::exceptionDetails).filterNotNull().toList()
		if (failedTasks2ElectricBoogaloo.isNotEmpty()) {
			errorsOccurred = true
			when (ui.showExceptions(failedTasks2ElectricBoogaloo, tasks.size, false)) {
				ExceptionListResult.CONTINUE -> {}
				ExceptionListResult.CANCEL -> cancelled = true
				ExceptionListResult.IGNORE -> cancelledStartGame = true
			}
		}
	}

	enum class ResolveResult {
		RETRY,
		QUIT,
		SUCCESS;
	}

	private fun validateAndResolve(nonFailedFirstTasks: List<DownloadTask>, clientHolder: ClientHolder): ResolveResult {
		ui.submitProgress(InstallProgress(Msgs.umValidatingExistingFiles()))

		// Validate existing files
		for (downloadTask in nonFailedFirstTasks.filter(DownloadTask::correctSide)) {
			downloadTask.validateExistingFile(opts.packFolder, clientHolder)
		}

		// Resolve CurseForge metadata
		val cfFiles = nonFailedFirstTasks.asSequence().filter { !it.alreadyUpToDate }
			.filter(DownloadTask::correctSide)
			.map { it.metadata }
			.filter { it.linkedFile != null }
			.filter { it.linkedFile!!.download.mode == DownloadMode.CURSEFORGE }.toList()
		if (cfFiles.isNotEmpty()) {
			ui.submitProgress(InstallProgress(Msgs.umResolvingCurseforgeMetadata()))
			val resolveFailures = resolveCfMetadata(cfFiles, opts.packFolder, clientHolder)
			if (resolveFailures.isNotEmpty()) {
				errorsOccurred = true
				return when (ui.showExceptions(resolveFailures, cfFiles.size, true)) {
					ExceptionListResult.CONTINUE -> {
						ResolveResult.RETRY
					}
					ExceptionListResult.CANCEL -> {
						cancelled = true
						ResolveResult.QUIT
					}
					ExceptionListResult.IGNORE -> {
						cancelledStartGame = true
						ResolveResult.QUIT
					}
				}
			}
		}
		return ResolveResult.SUCCESS
	}

	private fun showCancellationDialog() {
		when (ui.showCancellationDialog()) {
			CancellationResult.QUIT -> cancelled = true
			CancellationResult.CONTINUE -> cancelledStartGame = true
		}
	}

	// TODO: move to UI?
	private fun handleCancellation() {
		if (cancelled) {
			println(Msgs.umUpdateCancelledUser())
			exitProcess(1)
		} else if (cancelledStartGame) {
			println(Msgs.umUpdateCancelledUserStartGame())
			exitProcess(0)
		}
	}

}